### ISSUE TITLE
Babylon.js feedback on uniformity analysis.

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#2969]
TL;DR: few issues if `discard` means `demote_to_helper`.

Babylon.js analyzed the impact of the uniformity analysis on there shaders and found three places where it impacted them. There are few issues because Babylon.js is applying (an outdated AFAIK) guideline to avoid branches in shaders because they are slow, so they had few branches that caused non-uniform control flow. Below are the three cases that they found.

### Shadows

```rust
var uv = ...; // Compute the shadow UV based on position.
if uv.x < .00 || uv.x > 1.0 || uv.y < 0.0 || uv.y > 1.0 {
    return 1.0;
}
var shadowDepth = textureSample(shadowTexture, samp, uv).x;
return select(0.0, 1.0, pixelShadowDepth < shadowDepth);
```

The control flow after the `if` is non-uniform because it contains a `return` and `uv` is non-uniform. They can fix this easily by manually computing the UV LOD / gradients before the `if` and then use `textureSampleGrad`.

### Alpha test and early discard

All of Babylon.js's materials support alpha testing, which means that their fragment shaders will start with this snippet. It is doen really early so that blocks can be discarded before further texture fetches and lighting operations.

```rust
#ifdef ALPHATEST_AFTERALLALPHACOMPUTATIONS
if alpha < alphaCutOff {
    discard;
}
#endif
```

The control flow after the `if` is non-uniform because it contains a `discard` and `alpha` is non-uniform. Working around this issue is extremely tricky, because it would require computing all the gradient before the alpha test, consuming many registers and overall being complex to do due to Babylon.js's modular approach to shaders.

I previously brought this issue to the WGSL hours which spinned a bunch of discussion about adding back `demote_to_helper` (or replacing the current semantic of `discard` with demote). https://github.com/gpuweb/gpuweb/issues/921#issuecomment-1137658753 and https://github.com/gpuweb/gpuweb/issues/921#issuecomment-1137860604

### Clip planes and early discard

This is similar to the alpha test and has the exact same issue:

```rust
#ifdef CLIPPLANE
if fClipDistance > 0.0 {
    discard;
}
#endif
```

**Comment by kdashg** [comment]
I think the "Shadows" example is an acceptable level of work to ask for.
With resolution in #921 to have discard == demoteToHelper, the other examples should be satisfied also?
Closing if I got that right!

**Comment by sebavan** [comment]
You got it fully right :-) thanks a lot
